### ISSUE TITLE
How to understand environment step

### ISSUE THREAD

**Issue by tdr1991** [hill-a/stable-baselines#128]
Hi，
      I have create a custom environment named CE, each episode has n observations, if agent have read n observations the step function of CE will return done, that mean this episode is over. When I use a DQN agent, suppose current observation is obs[n-1] which last element of a episode,  agent call the step function of CE and get done, at the same time agent also call the reset function of CE, the reset function of CE will return first observation of next episode, at this train step agent look current observation and get a new observation that is the first observation of next episode, but these two observations are not in the same episode. Why? My understanding is if return done the current observation will reset to first observation of next episode. That is, there is no correlation between the two episodes.

**Comment by araffin** [comment]
Hello,
You problem is not super clear to me. Could you rephrase or elaborate your issue?

**Comment by doviettung96** [comment]
I think what he is asking is that:
next_state, reward, done, _ = env.step(action).
What is the meaning of next_state in case done == True? Is that the same as env.reset()? I have not looked at the docs in detail but I think normally we just throw that state away since done == True, we will break the loop, call state = env.reset().
About correlation, two episodes are independent. Each of them starts with env.reset() which is eventually a random function to set the state to a random value.
Does it help? 
Thanks.

**Comment by araffin** [comment]
>What is the meaning of next_state in case done == True? Is that the same as env.reset()? I have not looked at the docs in detail but I think normally we just throw that state away 

Not really, this correspond to the terminal observation. And yes, it is usually discarded. 
env.reset() returns the initial observation.

**Comment by araffin** [comment]
>Yes, in my work nest_state = env.reset() in case done==true.

This should not be the case, otherwise you don't respect the OpenAI Gym interface and RL Baselines may not work as expected.
